Sanitize option key for subscription

nagpai opened the pull request:
Replace spaces with underscores between words in option key names for sub products.

Fixes #4259 

#### Changes proposed in this Pull Request
Sanitize Subscription product option key name before they are saved in DB

Option keys, for example, tax names, are saved as `_wcpay_product_id_Tax name`  instead of `_wcpay_product_id_tax_name`. The PR replaces spaces with underscores, and makes the option name in all small case, before it gets concatenated and saved in the database. 


#### Testing instructions

* Turn taxes on for a WC store
* Add a new tax with a name containing spaces. For example ('Tax for climate')
* Apply the tax to a WC Pay subscriptions product (or leave it on all store products)
* Purchase the product with the tax applied to it
* Verify the wp_options table. There should be a new option called `_wcpay_product_id_test_tax_for_climate` (contains name of the tax with underscore instead of spaces)

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (Not applicable)


Post merge
- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : (QA Testing Not Applicable)
- [x] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable. - (Not applicable)
- [x] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable. - (Not applicable)

Commit:
Sanitize option key for subscription

Fix for #4259. Replace spaces with underscores between words in option key names for sub products.

elazzabi commented on includes/subscriptions/class-wc-payments-product-service.php:
```diff
@@ -128,11 +128,12 @@ public function get_wcpay_product_id( WC_Product $product, $test_mode = null ) :
 	public function get_wcpay_product_id_for_item( string $type ) : string {
-		if ( ! get_option( self::get_wcpay_product_id_option() . '_' . $type ) ) {
-			$this->create_product_for_item_type( $type );
+		$sanitized_type = str_replace( ' ', '_', trim( strtolower( $type ) ) );
```
This will effectively remove all spaces and replace them with `_`, I wonder if that's the only thing we want to do? ie, are there any other things we want to sanitize? Maybe `+`, `%` or other special characters...

Also, can we re-use some of the existing Woo functionality? The `sanitize_key()` function for example.

anu-rock replied:
A great start, @nagpai! :)

I was having thoughts similar to @elazzabi just now after looking at it more carefully. `sanitize_key` is purpose-built for scenarios like this and ensures no extraneous characters are present in the target string.

Going by its [source code](https://core.trac.wordpress.org/browser/tags/6.0/src/wp-includes/formatting.php#L2154), looks like it will remove even spaces. Although that might result in reduced readability compared to what you've done, it's a reasonable tradeoff.

nagpai replied:
Thank you @elazzabi and @anu-rock ! I had tried the `sanitize_key()` function earlier and passed that since it was removing the spaces and affecting reliability. Thanks to the reminder here about excluding any other special characters though! 

I think the solution here would be to first replace the spaces with underscores, and then run that through `sanitize_key`. I will test this and place for further review here 👍

anu-rock replied:
> I think the solution here would be to first replace the spaces with underscores

We may skip this entirely. Why do you think readability is important here?

nagpai replied:
> We may skip this entirely. Why do you think readability is important here?

My gut feel was if someone searches the db, they would look for `tax_with_spaces` instead of `taxwithspaces`. Coming to think of it if they look for `%spaces%` they will still find it. 

I missed reading your comment and for now I have passed the existing `$sanitized_type` formula through `sanitize_key()`.  Let me know if you would want me to keep it simple and move it to just `sanitize_key()` without bothering about underscores.

elazzabi reviewed (commented):
Thank you for working on this!

I left a comment about the sanitization process

nagpai committed:
Merge branch 'develop' into fix/4259-sanitize-tax-rate-name

nagpai reviewed (commented):
I accidentally clicked the `Update branch` button. It ended up pushing the commit. Sorry. It is blocked thankfully!

Commit:
Update sanitize code to remove special chars

The update will ensure the options name key does not have any special characters. Spaces replaced by underscore.

Commit:
Merge branch 'fix/4259-sanitize-tax-rate-name' of https://github.com/automattic/woocommerce-payments into fix/4259-sanitize-tax-rate-name

nagpai reviewed (commented):
On testing this commit, the tax name `Nagesh @#$%* Tax` resulted in a db option key name of `_wcpay_product_id_test_nagesh__tax`

kalessil commented on includes/subscriptions/class-wc-payments-product-service.php:
```diff
@@ -128,11 +128,12 @@ public function get_wcpay_product_id( WC_Product $product, $test_mode = null ) :
 	public function get_wcpay_product_id_for_item( string $type ) : string {
-		if ( ! get_option( self::get_wcpay_product_id_option() . '_' . $type ) ) {
-			$this->create_product_for_item_type( $type );
+		$sanitized_type = sanitize_key( str_replace( ' ', '_', trim( strtolower( $type ) ) ) );
```
A micro-optimisation suggession: `strtolower( trim( $type ) )`

nagpai replied:
Thanks for reviewing, and those very helpful early learnings for me. I have made the changes in the update 
https://github.com/Automattic/woocommerce-payments/pull/4496/commits/a891e1437fe8adf3adeb5ee77684d58acf24c453#

I have unit tests planned for this. Being the first time, I am taking a bit longer to bring it in :D

kalessil commented on includes/subscriptions/class-wc-payments-product-service.php:
```diff
@@ -128,11 +128,12 @@ public function get_wcpay_product_id( WC_Product $product, $test_mode = null ) :
 		}
 
-		return get_option( self::get_wcpay_product_id_option() . '_' . $type );
+		return get_option( self::get_wcpay_product_id_option() . '_' . $sanitized_type );
```
Also `self::get_wcpay_product_id_option() . '_' . $sanitized_type` can be a local variable - another micro optimization.

nagpai replied:
Thanks for sharing this opportunity to make the code cleaner following DRY! I saved this as a local variable `$option_key_name`. I hope this looks fine?

kalessil replied:
Yep, looks good!

kalessil reviewed (changes requested):
Yay, the first PR! I left a couple of hints + I'd love to see UTs introduced for this method. Thanks!

Commit:
Micro optimization

* Change sequence of `trim` function to help `strtolower` iterate over fewer characters

* Create local variable for repeated code

Commit:
Add unit test

Commit:
Code and test refactor

Moved sanitization code to its own static function. Added assertion to test the output of the sanitization function.

nagpai committed:
Merge branch 'develop' into fix/4259-sanitize-tax-rate-name

anu-rock commented on includes/subscriptions/class-wc-payments-product-service.php:
```diff
@@ -128,11 +128,22 @@ public function get_wcpay_product_id( WC_Product $product, $test_mode = null ) :
+	 * return string       Sanitized output.
+	 */
+	public static function sanitize_option_key( string $type ) {
+		return sanitize_key( str_replace( ' ', '_', strtolower( trim( $type ) ) ) );
```
`sanitize_key` will also [apply strtolower](https://github.com/WordPress/wordpress-develop/blob/6.0/src/wp-includes/formatting.php#L2158). We can remove our explicit use of it.

nagpai replied:
Thanks for sharing the reference! I will remove the `strtolower` from my line, since it is already taken care of by `sanitize_key`

nagpai replied:
Fixed here https://github.com/Automattic/woocommerce-payments/pull/4496/commits/9c3f75c0d55c429575811709930442123a648c34

anu-rock commented on tests/unit/subscriptions/test-class-wc-payments-product-service.php:
```diff
@@ -274,9 +274,7 @@ public function test_get_wcpay_product_id() {
 		$this->assertSame( $mock_product_id, $this->product_service->get_wcpay_product_id( $this->mock_product ) );
 	}
 
-	/**
```
Looks like this got removed accidentally?

nagpai replied:
Thanks for catching that! . I have added it back - https://github.com/Automattic/woocommerce-payments/pull/4496/commits/4ef7523e175471d595ff3fc949c865ebbdd04425

anu-rock commented on tests/unit/subscriptions/test-class-wc-payments-product-service.php:
```diff
@@ -286,13 +284,26 @@ public function test_get_wcpay_product_id_option() {
+			]
+		);
+
+		// If type is 'Test Tax *&^ name', the result should be _wcpay_product_id_live_test_tax__name.
```
I like this comment here 👍

nagpai replied:
Thank you :)

anu-rock commented on tests/unit/subscriptions/test-class-wc-payments-product-service.php:
```diff
@@ -286,13 +284,26 @@ public function test_get_wcpay_product_id_option() {
+		$test_type = 'Test Tax *&^ name';
+		$this->product_service->get_wcpay_product_id_for_item( $test_type );
+
+		$this->assertSame( 'test_tax__name', WC_Payments_Product_Service::sanitize_option_key( $test_type ) );
```
This should go in a separate test dedicated to `sanitize_option_key` :)

nagpai replied:
Thanks for the reference! I have fixed it here - https://github.com/Automattic/woocommerce-payments/pull/4496/commits/4ef7523e175471d595ff3fc949c865ebbdd04425

nagpai commented on includes/subscriptions/class-wc-payments-product-service.php:
```diff
@@ -128,11 +128,22 @@ public function get_wcpay_product_id( WC_Product $product, $test_mode = null ) :
+	 * Sanitize option key string to replace space with underscore, and remove special characters.
+	 *
+	 * @param string $type Non sanitized input.
+	 * return string       Sanitized output.
```
Ah. Looks like I copied it from [Line 128](https://github.com/automattic/woocommerce-payments/blob/083c42d6f2644fe3eee107e76f386931aff9a663/includes/subscriptions/class-wc-payments-product-service.php#L128). I will fix both. Thanks for catching this 🙏

nagpai replied:
Fixed here 
https://github.com/Automattic/woocommerce-payments/pull/4496/commits/9c3f75c0d55c429575811709930442123a648c34

anu-rock replied:
Thank you 👍

Commit:
Minor corrections and refactor

Commit:
Create separate test for sanitization method and add missing comment

kalessil commented on tests/unit/subscriptions/test-class-wc-payments-product-service.php:
```diff
@@ -286,13 +286,32 @@ public function test_get_wcpay_product_id_option() {
+	/**
+	 * Tests for WC_Payments_Product_Service::sanitize_option_key()
+	 */
+	public function test_sanitize_option_key() {
```
IMO we can drop this test case, as `test_get_wcpay_product_id_for_item` covers the logic.

nagpai replied:
@kalessil - I have removed the redundant test case, just now, here https://github.com/Automattic/woocommerce-payments/pull/4496/commits/e792fc0b78e0f64068513d65fe0e3ef4c8bc5d18

Let me know if this looks good to merge. Thanks :)

anu-rock reviewed (approved):
The revised code is looking all good. Cheers for your first approval 🎉 

There's an open comment from @kalessil regarding redundant test case. We may want to address it before merging the PR.

nagpai committed:
Merge branch 'develop' into fix/4259-sanitize-tax-rate-name

Commit:
Remove redundant test

Commit:
Merge branch 'fix/4259-sanitize-tax-rate-name' of https://github.com/automattic/woocommerce-payments into fix/4259-sanitize-tax-rate-name

kalessil reviewed (approved):
LGTM!
